src/VisualStudio/Core/Def/Implementation/ProjectSystem/DocumentProvider.cs

```diff
@@ -175,22 +206,115 @@ private ITextBuffer TryGetTextBufferFromDocData(object docData)
             }
         }
 
-        private void NewBufferOpened(uint docCookie, ITextBuffer textBuffer, DocumentKey documentKey, bool isCurrentContext)
+        private void NewBufferOpened(uint docCookie, ITextBuffer textBuffer, DocumentKey documentKey, bool isCurrentContext, CancellationToken cancellationToken)
         {
+            AssertIsForeground();
+
+            cancellationToken.ThrowIfCancellationRequested();
+
             StandardTextDocument document;
+            lock (_gate)
+            {
+                if (!_documentMap.TryGetValue(documentKey, out document))
+                {
+                    return;
+                }
+            }
 
-            if (_documentMap.TryGetValue(documentKey, out document))
+            document.ProcessOpen(textBuffer, isCurrentContext);
+
+            lock (_gate)
             {
-                document.ProcessOpen(textBuffer, isCurrentContext);
-                AddCookieOpenDocumentPair(docCookie, documentKey);
+                AddCookieOpenDocumentPair_NoLock(docCookie, documentKey);
             }
         }
 
         /// <summary>
         /// Notifies the document provider that this document is now registered in a project.
+        /// If we are on a foregroud thread, then this is done right away.
+        /// Otherwise, we schedule a task on foreground task scheduler.
         /// </summary>
         public void NotifyDocumentRegisteredToProject(IVisualStudioHostDocument document)
         {
+            if (IsForeground())
+            {
+                NotifyDocumentRegisteredToProject_Core(document, cancellationToken: CancellationToken.None);
+            }
+            else
+            {
+                var cts = new CancellationTokenSource();
+                var task = InvokeBelowInputPriority(() => NotifyDocumentRegisteredToProject_Core(document, cts.Token), cts.Token);
+                AddPendingDocumentInitializationTask(document, task, cts);
+            }
+        }
+
+        private void AddPendingDocumentInitializationTask(IVisualStudioHostDocument document, Task task, CancellationTokenSource cts)
+        {
+            var taskAndTokenSource = new TaskAndTokenSource() { Task = task, CancellationTokenSource = cts };
+            lock (_gate)
+            {
+                // Add taskAndTokenSource to the pending document initialization tasks.
+                // Check for cancellation before adding as the task might already have been completed/cancelled/faulted before we reached here.
+                if (!cts.IsCancellationRequested && !task.IsCompleted && !task.IsFaulted)
+                {
+                    _pendingDocumentInitializationTasks.Add(document.Id, taskAndTokenSource);
```

Comment:
We recently had some long discussions in the IDE team room about the reusability of Ids across documents. Should you just use the document object itself there?